### ISSUE TITLE
Display dose map colorbar in 2D view

### ISSUE THREAD

**Issue by cpinter** [SlicerRt/SlicerRT#51]
Right now, SlicerRT has the option to display colorbar for the isodose contours according to the Gy assigned to each one of them. However, when the dose map is displayed it would be also useful to see the association between colors in the map and the Gy values.

Important to consider that the colorbar must change when the window level is modified in RTDOSE volume display properties.

Migrated from https://app.assembla.com/spaces/slicerrt/tickets/650-display-dose-map-colorbar-in-2d-view/details

**Comment by cpinter** [comment]
Csaba: Good point, thanks for adding this. Please note that isodose colors and dose map colors are tied, so this is a bit harder than it seems.

**Comment by cpinter** [comment]
Ricardo A. Corredor:
![image](https://user-images.githubusercontent.com/1325980/30523816-3f6d09aa-9bb6-11e7-90dc-5c9c588fc766.png)

I realized two things about this ticket. First, the was already pointed out in #121. Second, Slicer can already display the scalar colorbar that adapts to changes in the window level (check the image I uploaded to this ticket Color Scalar Bar ).

Now, there are still some specific things that could be considered as issues or requirements:
- Adding a label to Slicer scalar colorbar with the units (e.g. set on top of the bar a text with the units (Gy) )
- There is no relation between the color table used for displaying the dose map and the color table used for displaying the isodoses. This could create some redundancy or confusion.

**Comment by cpinter** [comment]
2014-10-06 09:30

Csaba Pinter
"- There is no relation between the color table used for displaying the dose map and the color table used for displaying the isodoses. This could create some redundancy or confusion."
There is a relation, just not a dynamic one. Basically SlicerRT has an isodose color map that is loaded on startup. A dose color table is created from this isodose map when a dose volume is loaded. Unfortunately the two are not linked in a way that if one changes the other should too. This is covered in #315

2014-10-06 10:40

Ricardo A. Corredor
True, but maybe it is better to keep only this existing colorbar in Slicer, that is already dynamic, and adapt the isodose contour colors to those changes, am I wrong? why would you need the additional scalar colorbars for the isodose contours?

2014-10-06 10:41

Csaba Pinter
Because it has to be manually changeable in the isodose module

2014-10-06 11:18

Ricardo A. Corredor
Ok, but it is not easier/better to change directly the color lookup table assigned for dose data by manually setting custom colors to specific values? For example, that is what I usually do in ParaView where it is possible to set a color and a opacity to a value in the data range, via a colormap widget http://www.kitware.com/blog/home/post/573. In that way, the color settings would be automatically reflected both in the dose map and the isodose colors, keeping the same scalar colorbar. (Only questions to understand better, but you have more experience than me in the field :-) ) Thanks @pinter !!

2014-10-06 11:24

Csaba Pinter
Although the two color table nodes assign the same colors to the same levels, the isodose table has to be kept open for editing to the user who generates isodose surfaces. The dose volume color table is on the other hand contains a lot of color entries smoothing out these colors. The volume displayable manager can probably interpolate between the colors, so it is possible the isodose table can be used. What is for sure is that the current way is quite rigid and not even convenient. Unfortunately I cannot deal with this right now.

The best would be if you and @wangk came up with a specification-like document (Assembla wiki page) about the requirements that I and Andras also edit, and then we could come up with a design based on that.

2014-10-06 11:55

Ricardo A. Corredor
Sure @pinter ! My point of course is not to put you more work and push you more tickets :D It is specially to solve questions that help me to understand your decisions and hopefully part of these discussions would be useful (or not) to improve SlicerRT. I will try to see this point with @wangk, but at this moment it is still more important the beams module part. Thank you again!

2014-10-06 12:02

Csaba Pinter
I didn't take it like that, no worries :) I think such a rough specification would be useful to get the ball rolling, even if your focus is different for now, because apparently you would rely on it, and have definite requirements in that matter. This part of the code is something that needs overhaul anyway. Thanks!

2014-10-06 13:52

Kevin Wang
well. all are good points!

a bit background for this scalarbar issue. when I first started to develop isodose module, we needed a scalarbar that can show the label as Gy value(e.g. 30 Gy) instead of color value 12 which has no physical meaning. that is why we developed our own scalarbar and later on this code was integrated into the slicer core.

On another different project, some slicer developers have developed an enhanced scalarbar for dynamically updating according to current window/level setting, they based their implementation on the paraview's scalarbar.

There was some discussion on how to integrate the two scalarbar into one but things more slowly and it has not happened so far.

so in conclusion, you are right that we need features from both code. it just takes time.

**Comment by cpinter** [comment]
In the current version it is possible to show both the dose volume and isodose scalar bars:
![image](https://user-images.githubusercontent.com/1325980/62246928-b690bc00-b3b2-11e9-8672-14c9fe55c052.png)

The isodose color map is generated from the dose color map, so if the default window/leveling is not changed, the two are identical. In addition, both the isodose and the dose volume scalar bars are updated dynamically on changes in each, so their relationship can be seen side by side after the user decided to make adjustments.

Closing the ticket as fixed.
